client/checkout/blocks/upe-fields.js

```diff
@@ -0,0 +1,167 @@
+/**
+ * External dependencies
+ */
+import {
+	Elements,
+	ElementsConsumer,
+	PaymentElement,
+} from '@stripe/react-stripe-js';
+import { useEffect, useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import confirmUPEPayment from './confirm-upe-payment.js';
+import { getConfig } from 'utils/checkout';
+import { PAYMENT_METHOD_NAME_CARD } from '../constants.js';
+
+const WCPayUPEFields = ( {
+	api,
+	activePaymentMethod,
+	stripe,
+	elements,
+	billing: { billingData },
+	eventRegistration: {
+		onPaymentProcessing,
+		onCheckoutAfterProcessingWithSuccess,
+	},
+	emitResponse,
+} ) => {
+	const [ paymentIntentId, setPaymentIntentId ] = useState( null );
+	const [ clientSecret, setClientSecret ] = useState( null );
+	const [ hasRequestedIntent, setHasRequestedIntent ] = useState( false );
+	const [ isUPEComplete, setIsUPEComplete ] = useState( false );
+
+	const testMode = getConfig( 'testMode' );
+	const testCopy = (
+		<p>
+			<strong>Test mode:</strong> use the test VISA card 4242424242424242
+			with any expiry date and CVC.
+		</p>
+	);
+
+	useEffect( () => {
+		if ( paymentIntentId || hasRequestedIntent ) {
+			return;
+		}
+
+		async function createIntent() {
+			const response = await api.createIntent();
+			setPaymentIntentId( response.id );
+			setClientSecret( response.client_secret );
+		}
+		setHasRequestedIntent( true );
+
+		createIntent();
+	}, [ paymentIntentId, hasRequestedIntent, api ] );
+
+	// When it's time to process the payment, generate a Stripe payment method object.
+	useEffect(
+		() =>
+			onPaymentProcessing( () => {
+				if ( PAYMENT_METHOD_NAME_CARD !== activePaymentMethod ) {
+					return;
+				}
+
+				if ( ! isUPEComplete ) {
+					return {
+						type: 'error',
+						message: 'Your payment information is incomplete.',
+					};
+				}
+			} ),
+		// not sure if we need to disable this, but kept it as-is to ensure nothing breaks. Please consider passing all the deps.
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+		[ activePaymentMethod, isUPEComplete ]
+	);
+
+	// Once the server has completed payment processing, confirm the intent of necessary.
```

Comment:
```suggestion
	// Once the server has completed payment processing, confirm the intent if necessary.
```
I feel bad even putting this as a suggestion, but I've seen it and now I can't stop seeing it, sorry.